dashboard/modules/job/job_log_storage_client.py

```diff
@@ -13,45 +17,65 @@ class JobLogStorageClient:
     """
 
     # Number of last N lines to put in job message upon failure.
-    NUM_LOG_LINES_ON_ERROR = 10
-    # Maximum number of characters to print out of the logs to avoid
+    MAX_LOG_LINES_ON_ERROR = 10
+    # Max number of characters to print out of the logs to avoid
     # HUGE log outputs that bring down the api server
-    MAX_LOG_SIZE = 20000
+    MAX_LOG_SNIPPET_SIZE_ON_ERROR = 20000
+    # Max number of lines being fetched from log file per chunk (after
+    # every chunk event-loop has to be yielded to make sure that other tasks
+    # are not being starved)
+    MAX_LINES_PER_CHUNK = 10
 
-    def get_logs(self, job_id: str) -> str:
-        try:
-            with open(self.get_log_file_path(job_id), "r") as f:
-                return f.read()
-        except FileNotFoundError:
-            return ""
-
-    def tail_logs(self, job_id: str) -> Iterator[List[str]]:
-        return file_tail_iterator(self.get_log_file_path(job_id))
+    def tail_logs(
+        self,
+        job_id: str,
+        *,
+        max_lines_per_chunk: int = MAX_LINES_PER_CHUNK,
+    ) -> Iterator[List[str]]:
+        return file_tail_iterator(
+            self.get_log_file_path(job_id),
+            max_lines_per_chunk=max_lines_per_chunk,
+        )
 
-    def get_last_n_log_lines(
-        self, job_id: str, num_log_lines=NUM_LOG_LINES_ON_ERROR
+    async def get_logs(
+        self,
+        job_id: str,
+        *,
+        max_log_lines: int = INT32_MAX,  # unbounded
+        max_total_size: int = INT32_MAX,  # unbounded
     ) -> str:
         """
         Returns the last MAX_LOG_SIZE (20000) characters in the last
         `num_log_lines` lines.
 
         Args:
             job_id: The id of the job whose logs we want to return
-            num_log_lines: The number of lines to return.
+            max_log_lines: The number of lines to return.
         """
-        log_tail_iter = self.tail_logs(job_id)
-        log_tail_deque = deque(maxlen=num_log_lines)
-        for lines in log_tail_iter:
+        log_tail_deque = deque(maxlen=max_log_lines)
+
+        max_lines_per_chunk = min(self.MAX_LINES_PER_CHUNK, max_log_lines)
+        read_lines_count = 0
+
+        for lines in self.tail_logs(job_id, max_lines_per_chunk=max_lines_per_chunk):
```

Comment:
As I commented if we change the iterator to AsyncIterator this can be `async for` and we no longer need a `await asyncio.sleep(0)`.